packages/framework/aqueduct/src/data-object-factories/converterDataObjectFactory.ts

```diff
@@ -0,0 +1,193 @@
+	 * }
+	 * ```
+	 */
+	asyncGetDataForConversion: (root: ISharedDirectory) => Promise<TConversionData>;
+
+	/**
+	 * Convert the DataObject upon resolve (i.e. on retrieval of the DataStore).
+	 *
+	 * An example implementation could be changing which underlying DDS is used to represent the DataObject's data:
+	 * ```
+	 * (runtime, root, data) => {
+	 *     // ! These are not all real APIs and are simply used to convey the purpose of this method
+	 *     const mapContent = data.getContent();
+	 *     const newDirectory = SharedDirectory.create(runtime);
+	 *     newDirectory.populateContent(mapContent);
+	 *     root.set("directoryKey", newDirectory.handle);
+	 *     root.delete("mapKey");
+	 * }
+	 * ```
+	 * @param data - Provided by the "asyncGetDataForConversion" function
+	 */
+	convertDataObject: (
+		runtime: FluidDataStoreRuntime,
+		root: ISharedDirectory,
+		data: TConversionData,
+	) => void;
+
+	/**
+	 * If not provided, the Container will be closed after conversion due to underlying changes affecting the data model.
+	 */
+	refreshDataObject?: () => Promise<void>;
+}
+
+/**
+ * ConverterDataObjectFactory is the IFluidDataStoreFactory for converting DataObjects.
+ * See ConverterDataObjectFactoryProps for more information on how to utilize this factory.
+ *
+ * @internal
+ */
+export class ConverterDataObjectFactory<
+	TObj extends DataObject<I>,
+	TConversionData,
+	I extends DataObjectTypes = DataObjectTypes,
+> extends DataObjectFactory<TObj, I> {
+	private convertLock = false;
+
+	// ! TODO: add new DataStoreMessageType.Conversion
+	private static readonly conversionContent = "conversion";
+
+	public constructor(props: ConverterDataObjectFactoryProps<TObj, TConversionData, I>) {
+		const submitConversionOp = (runtime: FluidDataStoreRuntime): void => {
+			runtime.submitMessage(
+				DataStoreMessageType.ChannelOp,
+				ConverterDataObjectFactory.conversionContent,
+				undefined,
+			);
+		};
+
+		const fullConvertDataStore = async (runtime: IFluidDataStoreChannel): Promise<void> => {
+			const realRuntime = runtime as FluidDataStoreRuntime;
+			const root = (await realRuntime.getChannel(
+				dataObjectRootDirectoryId,
+			)) as ISharedDirectory;
+			if (!this.convertLock && (await props.isConversionNeeded(root))) {
+				this.convertLock = true;
+				const data = await props.asyncGetDataForConversion(root);
+
+				// ! TODO: ensure these ops aren't sent immediately AB#41625
+				submitConversionOp(realRuntime);
+				props.convertDataObject(realRuntime, root, data);
+				this.convertLock = false;
+			}
+		};
+
+		const runtimeClass = props.runtimeClass ?? FluidDataStoreRuntime;
+
+		super({
+			...props,
+			afterBindRuntime: fullConvertDataStore,
+			runtimeClass: class ConverterDataStoreRuntime extends runtimeClass {
+				private conversionOpSeqNum = -1;
+				private readonly seqNumsToSkip = new Set<number>();
+
+				public processMessages(messageCollection: IRuntimeMessageCollection): void {
```

Comment:
i feel like the converter itself should really be in its own review. this is adding persisted state, and has back and cross compat concerns. i won't be around to follow the review, but this probably needs a design, and should have a lot more testing, since there is none currently. some good people to review this with would @Abe27342 , @agarwal-navin, @markfields and probably others.

Reply:
I agree with doing a proper design review on the larger effort. As the description points out, this is only an intermediate step of the larger feature [AB#41618](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/41618), and with this PR the `ConverterDataObjectFactory` is not used yet of production scenarios. So for me the main question is if this PR should be blocked by the larger review for the feature design or not.

Reply:
i'm also a little worried about the other change here, more that you may find they need to change, or you need other capabilities. using a test branch lets you iterate quickly without accruing  debt.